distribution/tools/ansi-console/src/main/java/org/elasticsearch/io/ansi/AnsiConsoleLoader.java

```diff
@@ -7,33 +7,72 @@
  */
 package org.elasticsearch.io.ansi;
 
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.bootstrap.ConsoleLoader;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.SuppressForbidden;
+import org.fusesource.jansi.Ansi;
 import org.fusesource.jansi.AnsiConsole;
 import org.fusesource.jansi.AnsiPrintStream;
 import org.fusesource.jansi.AnsiType;
+import org.fusesource.jansi.io.AnsiOutputStream;
 
-import java.io.PrintStream;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.charset.Charset;
 import java.util.function.Supplier;
 
+import static org.apache.logging.log4j.LogManager.getLogger;
+
 /**
- * Loads the({@link PrintStream} print stream) from {@link AnsiConsole} and checks whether it meets our requirements for a "Console".
+ * Loads the {@link AnsiConsole} and checks whether it meets our requirements for a "Console".
  * @see org.elasticsearch.bootstrap.ConsoleLoader
  */
-public class AnsiConsoleLoader implements Supplier<PrintStream> {
+public class AnsiConsoleLoader implements Supplier<ConsoleLoader.Console> {
+
+    private static final Logger logger = getLogger(AnsiConsoleLoader.class);
 
-    public PrintStream get() {
+    public ConsoleLoader.Console get() {
         final AnsiPrintStream out = AnsiConsole.out();
+        return newConsole(out);
+    }
+
+    // package-private for tests
+    static @Nullable ConsoleLoader.Console newConsole(AnsiPrintStream out) {
         if (isValidConsole(out)) {
-            return out;
+            return new ConsoleLoader.Console(out, () -> out.getTerminalWidth(), Ansi.isEnabled(), tryExtractPrintCharset(out));
         } else {
             return null;
         }
     }
 
-    static boolean isValidConsole(AnsiPrintStream out) {
+    private static boolean isValidConsole(AnsiPrintStream out) {
         return out != null // cannot load stdout
             && out.getType() != AnsiType.Redirected // output is a pipe (etc)
             && out.getType() != AnsiType.Unsupported // could not determine terminal type
             && out.getTerminalWidth() > 0 // docker, non-terminal logs
         ;
     }
+
+    /**
+     * Uses reflection on the JANSI lib in order to expose the {@code Charset} used to encode the console's print stream.
+     * The {@code Charset} is not otherwise exposed by the library, and this avoids replicating the charset selection logic in out code.
+     */
+    @SuppressForbidden(reason = "Best effort exposing print stream's charset with reflection")
+    @Nullable
+    private static Charset tryExtractPrintCharset(AnsiPrintStream ansiPrintStream) {
```

Comment:
I don't love the fact that we have to use reflection either but I think that given that
- There is no way to override the method and field, so there is no risk of some injection attack of sorts 
- We don't do automatic dependency upgrades so that we can catch any changes that would trigger an error here
- The character set being very handy for a user friendly output (????????? :-1:  )

++ :+1: